bin/pub_insights.dart

```diff
@@ -36,8 +81,89 @@ Future<void> main(List<String> arguments) async {
 
   final Stopwatch stopwatch = Stopwatch()..start();
 
-  print('Discovering packages...');
-  final packages = await client.listPackages();
+  final mode = argResults.option(_fetchModeFlag) ?? '';
+  print('fetch mode "$mode"...');
+  switch (mode) {
+    case _fetchModeAll:
+      print('Discovering all packages...');
+      final packages = await client.listPackages();
+      await downloadPackages(packages, stopwatch,
+          metadataDir: metadataDir,
+          versionsDir: versionsDir,
+          packageDir: packageDir,
+          entriesDir: entriesDir,
+          scoresDir: scoresDir,
+          tablesDir: tablesDir);
+      break;
+    case _fetchModePopular:
+      print('Discovering popular packages...');
+      final packages = await client.listPopularPackages();
+      await downloadPackages(packages, stopwatch,
+          metadataDir: metadataDir,
+          versionsDir: versionsDir,
+          packageDir: packageDir,
+          entriesDir: entriesDir,
+          scoresDir: scoresDir,
+          tablesDir: tablesDir);
+      break;
+    case _fetchModeFile:
+      if (!argResults.wasParsed(_fetchFileFlag)) {
+        print('File flag must be set when using $_fetchModeFile');
+        exit(1);
+      }
+      final fetchFile = argResults.option(_fetchFileFlag) as String;
+      print('Discovering packages from $fetchFile...');
+      final packages = await client.listPackagesFromFile(path.relative(fetchFile));
+      await downloadPackages(packages, stopwatch,
+          metadataDir: metadataDir,
+          versionsDir: versionsDir,
+          packageDir: packageDir,
+          entriesDir: entriesDir,
+          scoresDir: scoresDir,
+          tablesDir: tablesDir);
+      break;
+    case _fetchModeNone:
+      print('Skipping downloading...');
+      break;
+    default:
+      print('Unknown fetch mode $mode');
+      break;
+  }
+  stopwatch.reset();
+
+  if (argResults.flag(_optimizeTablesFlag)) {
+    print('Optimizing tables ...');
+    final reports = {
+      'package_versions.json': versionsDir,
+      'package_scores.json': scoresDir,
```

Comment:
I removed the package archive entries table as it takes a while to regenerate. Now that we can skip this optimizing step, consider re-adding this table:

```suggestion
      'package_versions.json': versionsDir,
      'package_scores.json': scoresDir,
      'package_archive_entries.json': entriesDir,
```

Reply:
I would like to consider new data out of scope for this pr. It already has a weird collection of changes.

Reply:
Sounds good